api/envoy/config/core/v3/protocol.proto

```diff
@@ -100,6 +100,15 @@ message HttpProtocolOptions {
   // If this setting is not specified, the value defaults to ALLOW.
   // Note: upstream responses are not affected by this setting.
   HeadersWithUnderscoresAction headers_with_underscores_action = 5;
+
+  // If true, the upstreams will attempt to negotiate ALPN based on the protocol selected by
+  // :ref:`protocol_selection <envoy_api_field_config.cluster.v3.Cluster.protocol_selection>`.
+  // The following protocol to ALPN mapping will be used:
+  //   HTTP/1.0 -> http/1.0
+  //   HTTP/1.1 -> http/1.1
+  //   HTTP/2   -> h2
+  //   HTTP/3   -> h3
+  bool upstream_alpn_from_selected_protocol = 6;
```

Comment:
I think being pedantic, ALPN is a transport attribute. Should this either be in the Cluster config (next to protocol selection) or in the TLS config? It's a bit asymmetrical to put it in HTTP protocol options. Otherwise, LGTM.

Reply:
I'm confused. If I read documentation correctly this is the behavior for the `protocol_selection == USE_CONFIGURED_PROTOCOL`, am I right?

Reply:
`protocol_selection` determines which connection pool/codec is selected, but it does not impact the ALPN protocols that Envoy tries to negotiate, so if your cluster is configured to negotiate `h2,http/1.1` but you selected HTTP/1.1 as the protocol your upstream is likely going to blow up because it negotiated HTTP/2 but you're sending HTTP/1.1.

Reply:
Hmm, apologies for the neophyte question as I do not quite understand the hierarchy in cluster manager. Does the connection pool have hosts with single protocol only (i.e. H2) or can a single pool have hosts with a mix of H1, H2 and H3? If it is only one protocol shouldn't it define what is negotiated over ALPN?

Reply:
Each conn pool today only handles one protocol. We could have the connection pools be responsible for altering the ALPN, I don't think it changes this PR significantly (we'd just move the logic to wrap the socket options to the conn pool ctor instead of the cluster manager). 

I don't think we can just pin the ALPN based on protocol selection by default, since there are use case out in the wild that use custom ALPNs, e.g. Istio uses `istio-h2` instead of `h2` at times. One option here would be to specify a protocol -> ALPN mapping in config, which would then allow us to just have the connection pools use that to configure the ALPN. WDYT?

Reply:
This is a really common point of misconfiguration and I do really like the idea of just doing the right thing by default and otherwise making people override it (like in the Istio case). Is there a way we can do that? I'm not sure we can without a deprecation cycle though?

Either way I agree with @htuch that this feels like something that belongs with protocol selection, either as an additional boolean, or possibly even an enum variant? USE_DOWNSTREAM_PROTOCOL_AND_SET_ALPN? Or maybe setting ALPN can be the default for the use downstream protocol variant as otherwise it makes little sense? WDYT?

Reply:
Yes, I think it is a good idea. Just to make sure I really understand, if transport_socket has alpn_protocols specified we would use these in TLS negotiation. If it is not specified then it would be determined based on the connection pool protocol.

Reply:
> Just to make sure I really understand, if transport_socket has alpn_protocols specified we would use these in TLS negotiation. If it is not specified then it would be determined based on the connection pool protocol.

I think that would be one way of doing it, yes. I think we need to think through back-compat though to make sure we don't break anyone's existing deployment without a deprecate/warn cycle (if needed).

Reply:
I like the idea of having the upstream TLS context value take precedence if set for non-standard configuration while doing the right thing by default, I think that would make configuring ALPN much easier for most people.

Regarding backwards compat, there are currently two ways that ALPN can be set:

1) configuring it as part of the upstream TLS context
2) provided to the LB as part of the TransportSocketOptions provided by the LBContext

to not change existing behavior, I think we'd want a ALPN selection algorithm that looks something like this:

if ALPN set in TransportSocketOptions, use that
if ALPN is not set on the upstream TLS context, use the protocol specific one
otherwise use upstream TLS context

I think coupling this with the protocol selection option makes sense. On the choice between adding an enum or changing the default, I think the only concern I'd have around changing the default would be if people are intentionally not setting the ALPN for whatever reason. Changing the default would then make ALPN negotiation happen where it wasn't being done previously, which would be a change in behavior. Given that I'm leaning to making this a new option.

Does this sound reasonable to everyone?

Reply:
Yes this sounds like a great plan, especially no new config. I would recommend putting the behavior behind a feature flag with default on and if anyone complains we can deal with it then.

Reply:
I think @snowp is suggesting we do have a protocol selection config field. Are you saying yes to that @mattklein123 or yes to changing the default with runtime guard?

Reply:
AIUI @snowp is suggesting to not have any new config and to go through a set of steps to figure out which ALPN to use. This sounds right to me and should be safe for all users. The feature flag would allow us to turn it off if that is not the case and revisit. @snowp please correct me if I am misunderstanding.

Reply:
I was suggesting adding another enum, mostly because I don't feel that confident that runtime flags are sufficient given the number of Envoy deployments I know about that don't have any runtime infra set up. The "use protocol specific ALPN" step was to be gated on this enum value in my mind. 

That said, if we feel that a runtime guard is sufficient and in line with our policy on breaking changes I could go either way on this. This seems like it should be the default and would simplify configuration a lot, I'm just concerned about the potential fallout of a breaking change.

Reply:
> That said, if we feel that a runtime guard is sufficient and in line with our policy on breaking changes I could go either way on this. This seems like it should be the default and would simplify configuration a lot, I'm just concerned about the potential fallout of a breaking change.

Runtime can be set using bootstrap/CLI, so IMO it can be used by anyone regardless of their infra, so I think it's fine for this purpose (and we have assumed this for other feature flagging). I would be fine adding no config for this and just doing it if we have a guard mainly because I think we have an opportunity here to clean up a major usability issue.